init.lua

```diff
@@ -1,77 +1,302 @@
+local mod_name = "wielded_light"
+
 local update_interval = 0.2
+local cleanup_interval = update_interval*3
+local velocity_projection = update_interval * 1
 local level_delta = 2
+local max_light_level = 14
 local shiny_items = {}
 
---- Shining API ---
-wielded_light = {}
+local active_lights = {}
+local light_recalcs = {}
+local tracked_entities = {}
+local update_callbacks = {}
+local update_player_callbacks = {}
 
-function wielded_light.update_light(pos, light_level)
-	local around_vector = {
-			{x=0, y=0, z=0},
-			{x=0, y=1, z=0}, {x=0, y=-1, z=0},
-			{x=1, y=0, z=0}, {x=-1, y=0, z=0},
-			{x=0, y=0, z=1}, {x=0, y=0, z=1},
-		}
-	local do_update = false
-	local old_value = 0
-	local timer
-	local light_pos
-	for _, around in ipairs(around_vector) do
-		light_pos = vector.add(pos, around)
-		local name = minetest.get_node(light_pos).name
-		if name == "air" and (minetest.get_node_light(light_pos) or 0) < light_level then
-			do_update = true
-			break
-		elseif name:sub(1,13) == "wielded_light" then -- Update existing light node and timer
-			old_value = tonumber(name:sub(15))
-			if light_level > old_value then
-				do_update = true
-			else
-				timer = minetest.get_node_timer(light_pos)
-				local elapsed = timer:get_elapsed()
-				if elapsed > (update_interval * 1.5) then
-					-- The timer is set to 3x update_interval
-					-- This node was not updated the last interval and may
-					-- is disabled before the next step
-					-- Therefore the light should be re-set to avoid flicker
-					do_update = true
+local cat_id = 0
+local cat_codes = {}
+local function get_cat_code(cat)
+	if not cat_codes[cat] then
+		if cat_id >= 256 then
+			error("Wielded item category limit exceeded, maximum 256 wield categories")
+		end
+		local code = string.format("%02x", cat_id)
+		cat_id = cat_id+1
+		cat_codes[cat] = code
+	end
+	return cat_codes[cat]
+end
+
+local function entity_still_exists(entity)
+	return entity and (entity.obj:is_player() or entity.obj:get_entity_name() or false)
+end
+
+local function entity_pos(userdata, offset)
```

Comment:
You could also call `userdata` -> `obj` or `entity` to clarify that it's an ObjectRef. Same for the other functions, where applicable.

Reply:
Yeah I've renamed that to `obj`, I think in this code we should be using `entity` to refer to the custom object made for tracking and `obj` for the actual userdata / ObjectRef being tracked